docs/design/features/OsrDetailsAndDebugging.md

```diff
@@ -0,0 +1,382 @@
+hit limit threshold. If a second thread arrives while the OSR method
+is being created it will simply have its counter reloaded and continue
+executing Tier0 code for the time being. Eventually it may come back
+
+Once the method is ready the initiating thread is transitioned to the
+OSR method; any subsequently arriving executions from other active
+frames will likewise transition as they hit the patchpoint.
+
+The OSR method does not return control to the Tier0 method (it is a
+"full continuation" of the Tier0 method at the patchpoint).
+
+### Creation of the OSR Method
+
+The OSR method is specific to a Tier0 method at a specific IL
+offset. When the runtime decides to create an OSR method it invokes
+the jit and passes a special OSR flag and the IL offset.
+
+This compilation is similar to a normal optimized compilation, with 
+a few twists:
+* Importation starts at the specified IL offset and pulls in all the
+code reachable from that point. Typically (but not always) the method entry (IL offset 0) is unreachable
+and so the OSR method only imports a subset of the full method.
+* Control normally branches from fgFirstBB to the BB at the OSR IL offset (the "osr entry").
+Special care is needed when the osr entry is in the middle of a try or in a nested try.
+* If dynamic PGO is enabled the OSR method will read the PGO data captured by the Tier0
+method, but we also instrument the OSR method to ensure any Tier1 version we produce
+later on sees all the PGO data it would have seen if we'd forcibly kept the method at Tier0.
+
+The JIT may need to import the entry if it can be reached by tail
+recursion to loop optimizations, even if there no explicit IL branch
+to offset 0 in the IL reachable from the OSR Il offset. This was the
+source of a number of subtle bugs -- the call site that eventually may
+loop back to the method entry could come from an inlinee, or via
+devirtualization, etc.
+
+For the most part the jit treats an OSR method just like any other
+optimized jitted method. But aside from importation, frame layout,
+pgo, and prolog/epilog codegen details bleed through in only a handful
+of places.
+
+#### Frame Layout
+
+(forthcoming)
+
+#### OSR Prolog
+
+(forthcoming)
+
+#### OSR Epilog
+
+(forthcoming)
+
+#### Funclets in OSR Methods
+
+(forthcoming)
+
+### Execution of an OSR Method
+
+OSR methods are never called directly; they can only be invoked by `CORINFO_HELP_PATCHPOINT` when called from a Tier0 method with patchpoints.
+
+When the OSR method returns, it cleans up both its own stack and the
+Tier0 method stack.
+
+Note if a Tier0 method is recursive and has loops there can be some interesting dynamics. After a sufficient amount of looping an OSR method will be created, and the currently active Tier0 instance will transition to the OSR method. When th OSR method makes a recursive call, it will invoke the Tier0 method, which will then fairly quickly transition to the OSR version just created.
```

Comment:
```suggestion
Note if a Tier0 method is recursive and has loops there can be some interesting dynamics. After a sufficient amount of looping an OSR method will be created, and the currently active Tier0 instance will transition to the OSR method. When the OSR method makes a recursive call, it will invoke the Tier0 method, which will then fairly quickly transition to the OSR version just created.
```